reconcile/openshift_rolebindings.py

```diff
@@ -19,40 +31,194 @@
+    openshift_service_accounts: list[ServiceAccountSpec]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def get_users_desired_state(self) -> list[dict[str, str]]:
+        return [
+            {"cluster": self.cluster.name, "user": username}
+            for username in self.username_list
+        ]
+
+    @staticmethod
+    def create_role_binding_spec(
+        access: AccessV1,
+        oc_map: ob.ClusterMap | None = None,
+        users: list[UserV1] | None = None,
+        enforced_user_keys: list[str] | None = None,
+        bots: list[BotV1] | None = None,
+    ) -> Optional["RoleBindingSpec"]:
+        if not access.namespace:
+            return None
+        if not (access.role or access.cluster_role):
+            return None
+        privileged = access.namespace.cluster_admin or False
+        if oc_map and not oc_map.get(access.namespace.cluster.name):
+            return None
+        auth_dict = [auth.dict(by_alias=True) for auth in access.namespace.cluster.auth]
+        username_list = RoleBindingSpec.get_usernames_from_role(
+            users,
+            ob.determine_user_keys_for_access(
+                access.namespace.cluster.name,
+                auth_dict,
+                enforced_user_keys,
+            ),
+        )
+        service_accounts = ServiceAccountSpec.create_sa_spec(bots) if bots else []
+        return RoleBindingSpec(
+            role_name=access.role or access.cluster_role,
+            role_kind="Role" if access.role else "ClusterRole",
+            namespace=access.namespace,
+            cluster=access.namespace.cluster,
+            privileged=privileged,
+            username_list=username_list,
+            openshift_service_accounts=service_accounts,
+        )
+
+    @staticmethod
+    def create_rb_specs_from_role(
+        role: RoleV1,
+        oc_map: ob.ClusterMap | None = None,
+        enforced_user_keys: list[str] | None = None,
+    ) -> list["RoleBindingSpec"]:
+        rolebinding_spec_list = [
+            role_binding_spec
+            for access in role.access or []
+            if (
+                access.namespace
+                and is_valid_namespace(access.namespace)
+                and (
+                    role_binding_spec := RoleBindingSpec.create_role_binding_spec(
+                        access, oc_map, role.users, enforced_user_keys, role.bots
+                    )
+                )
+            )
+        ]
+        return rolebinding_spec_list
+
+    @staticmethod
+    def get_usernames_from_role(
+        users: list[UserV1] | None = None, user_keys: list[str] | None = None
+    ) -> set[str]:
+        return {
+            getattr(user, user_key)
+            for user in users or []
+            for user_key in user_keys or []
+            if hasattr(user, user_key)
         }
-      }
-      role
-    }
-    expirationDate
-  }
-}
-"""
 
+    def construct_user_oc_resource(
+        self, user: str, force_cluster_role_ref: bool
```

Comment:
do we need use `force_cluster_role_ref` in so many places, how about just use it once during `RoleBindingSpec` instance creation?

```python
role_kind="ClusterRole" if force_cluster_role_ref or access.cluster_role else "Role",
```

Reply:
to support backward compatibility deployment, let's change `FORCE_CLUSTER_ROLE_REF` to an input param for `run` func, maybe reverse it to `--support-role-ref`, default `False`.

so the role_kind can be

```python
        role_kind = "Role" if access.role and support_role_ref else "ClusterRole"
```

* promote this PR
* migrate data
* specify `--support-role-ref` in deployment
* new PR to remove this param